🎉 New Source: GCS

tuanchris opened the pull request:
## What
Add new source GCS. This connecter will: 
- Supports service account for authentication
- Get all files from a directory
- Scan for CSV files 
- Generate streams for each of the CSV files 
- Read the CSV files
[#11135](https://github.com/airbytehq/airbyte/issues/11135) @YowanR 

![image](https://user-images.githubusercontent.com/52090179/219542305-fe90df77-9d1c-4d32-bfd4-e01076bdfd5a.png)

## How
- Authenticate using `google-cloud-storage` database and a service account
-  Create a GCS client, a bucket object, and use `get_blobs` method to list all blobs
- Filter `.csv` files
- Read the first 0.1 mb of the file to create a json schema object 
- Use pandas to read the files to memory and write to AirbyteStreams

## Future improvements to be made: 
- Add different authentication method
- Support file compression
- Support other file formats: JSON, Parquet, etc. 
- Handle bigger files size 

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [x] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [x] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

YowanR commented:
@sh4sh @natalyjazzviolin can you please take a look? 🙏

marcosmarxm reviewed (changes requested):
Hello @tuanchris 👋 thanks for the contribution. I tried to test the connector. It is working reading one file but not a folder.

In my case I created the folder `test_folder` and added two csv files.
<img width="434" alt="image" src="https://user-images.githubusercontent.com/5154322/221541672-075d7684-404e-45e6-bbe7-d7ad1810c13c.png">


The config below doesn't work:
```
{
    "gcs_path": "test_folder/",
    "gcs_bucket": "airbyte-integration-test-source-gcs",
    "service_account": "..."
  }
```

Also it is missing the documentation and instructions to setup the connector.

Today the connector it isn't different from the File using GCS option. Probably if the connector work reading a folder it can be accepted, probably would be better a strategy to read incremental.

marcosmarxm commented on airbyte-integrations/connectors/source-gcs/source_gcs/source.py:
```diff
@@ -0,0 +1,69 @@
+        blobs = get_gcs_blobs(config)
+        for blob in blobs:
+            # Read the first 0.1MB of the file to determine schema
+            df = read_csv_file(blob, limit_bytes=102400)
```
Can you create a meaningful variable for the value 102400? `MIN_BYTES_READ_GENERATE_SCHEMA`

tuanchris commented:
Thanks @marcosmarxm for the review. I have: 
- Updated the logic of `get_gcs_blobs` to filter only `.csv` files
- Added docs
The connector should work with multiple files in GCS now. Here's my gcs structure:

![image](https://user-images.githubusercontent.com/52090179/221969251-65b2d426-422b-49a5-8861-172d6b058fea.png)

And here's the discovery results: 

```json
{
  "type": "CATALOG",
  "catalog": {
    "streams": [
      {
        "name": "film",
        "json_schema": {
          "$schema": "http://json-schema.org/draft-07/schema#",
          "type": "object",
          "properties": {
            "film_id": { "type": "string" },
            "title": { "type": "string" },
            "release_year": { "type": "string" },
            "language_id": { "type": "string" },
            "rental_duration": { "type": "string" },
            "rental_rate": { "type": "string" },
            "replacement_cost": { "type": "string" },
            "rating": { "type": "string" }
          }
        },
        "supported_sync_modes": ["full_refresh"]
      },
      {
        "name": "actor",
        "json_schema": {
          "$schema": "http://json-schema.org/draft-07/schema#",
          "type": "object",
          "properties": {
            "actor_id": { "type": "string" },
            "first_name": { "type": "string" },
            "last_name": { "type": "string" }
          }
        },
        "supported_sync_modes": ["full_refresh"]
      }
    ]
  }
}
```

tuanchris commented on airbyte-integrations/connectors/source-gcs/setup.py:
```diff
@@ -0,0 +1,29 @@
+
+from setuptools import find_packages, setup
+
+MAIN_REQUIREMENTS = [
```
@sh4sh I was having this issue installing pandas as a requirement https://github.com/airbytehq/airbyte/issues/12760. I had to use a different base image with pandas installed already https://github.com/airbytehq/airbyte/blob/105f4d2423756fadd35eafb4ae6accb7889f1ff3/airbyte-integrations/connectors/source-gcs/Dockerfile#L1

davydov-d commented on airbyte-integrations/connectors/source-gcs/Dockerfile:
```diff
@@ -0,0 +1,38 @@
+FROM nickgryg/alpine-pandas:3.9.12 as base
```
Could you please try adding `RUN apt-get install lib32stdc++6` to Dockerfile instead of using a custom base image? (and adding pandas to setup.py)

tuanchris replied:
I tried adding this: `apk --no-cache add libstdc++`, `apk --no-cache add libstdc++6` and `lib32stdc++6`  
but got the following error still: 
```
Traceback (most recent call last):
  File "/airbyte/integration_code/main.py", line 9, in <module>
    from source_gcs import SourceGCS
  File "/airbyte/integration_code/source_gcs/__init__.py", line 6, in <module>
    from .source import SourceGCS
  File "/airbyte/integration_code/source_gcs/source.py", line 23, in <module>
    from .helpers import construct_file_schema, get_gcs_blobs, read_csv_file, get_stream_name
  File "/airbyte/integration_code/source_gcs/helpers.py", line 8, in <module>
    import pandas as pd
  File "/usr/local/lib/python3.9/site-packages/pandas/__init__.py", line 48, in <module>
    from pandas.core.api import (
  File "/usr/local/lib/python3.9/site-packages/pandas/core/api.py", line 47, in <module>
    from pandas.core.groupby import (
  File "/usr/local/lib/python3.9/site-packages/pandas/core/groupby/__init__.py", line 1, in <module>
    from pandas.core.groupby.generic import (
  File "/usr/local/lib/python3.9/site-packages/pandas/core/groupby/generic.py", line 76, in <module>
    from pandas.core.frame import DataFrame
  File "/usr/local/lib/python3.9/site-packages/pandas/core/frame.py", line 172, in <module>
    from pandas.core.generic import NDFrame
  File "/usr/local/lib/python3.9/site-packages/pandas/core/generic.py", line 169, in <module>
    from pandas.core.window import (
  File "/usr/local/lib/python3.9/site-packages/pandas/core/window/__init__.py", line 1, in <module>
    from pandas.core.window.ewm import (
  File "/usr/local/lib/python3.9/site-packages/pandas/core/window/ewm.py", line 15, in <module>
    import pandas._libs.window.aggregations as window_aggregations
ImportError: Error loading shared library libstdc++.so.6: No such file or directory (needed by /usr/local/lib/python3.9/site-packages/pandas/_libs/window/aggregations.cpython-39-aarch64-linux-gnu.so)
```

davydov-d replied:
I have found a solution! Please use `slim` image instead of `alpine`, just like [here](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-salesforce/Dockerfile#L1)

davydov-d commented on airbyte-integrations/connectors/source-gcs/source_gcs/helpers.py:
```diff
@@ -0,0 +1,57 @@
+    schema = {
+        "$schema": "http://json-schema.org/draft-07/schema#",
+        "type": "object",
+        "properties": {col: {"type": column_types[col]} for col in column_types},
```
```suggestion
        "properties": {col: {"type": "string"} for col in df.columns},
```

davydov-d commented on airbyte-integrations/connectors/source-gcs/source_gcs/helpers.py:
```diff
@@ -0,0 +1,57 @@
+
+def construct_file_schema(df):
+    # Fix all columns to string for maximum compability
+    column_types = {col: "string" for col in df.columns}
```
```suggestion
```

davydov-d commented on airbyte-integrations/connectors/source-gcs/source_gcs/source.py:
```diff
@@ -0,0 +1,71 @@
+        for blob in blobs:
+            # Read the first 0.1MB of the file to determine schema
+            df = read_csv_file(
+                blob, limit_bytes=MIN_BYTES_READ_GENERATE_SCHEMA)
```
why do you need to read 0.1 MB if you declare all columns to be strings? I believe reading just a single line would be enough to get the column names

tuanchris replied:
I don't think the gcs api allows us to read a single line of a CSV. Therefore I was thinking 0.1 MB should be enough to include all column names

davydov-d replied:
According to this [doc](https://cloud.google.com/storage/docs/streaming-downloads#storage-stream-download-object-python), the client library exposes a `download_blob_to_stream` method which returns a file like object. So after you get the file like object you can do something like 
```
for line in file_like_obj:
  first_line = line
  break
```

davydov-d commented on airbyte-integrations/connectors/source-gcs/source_gcs/source.py:
```diff
@@ -0,0 +1,71 @@
+        return AirbyteCatalog(streams=streams)
+
+    def read(
+        self, logger: AirbyteLogger, config: json, catalog: ConfiguredAirbyteCatalog, state: Dict[str, any]
```
you're not using catalog at all! this means a user won't be able to turn off streams he doesn't need. Please fix this

tuanchris replied:
@davydov-d Do you have an example of a catalog implementation for an undefined source like this?

davydov-d replied:
I have a json representation of a sample configured catalog which looks as follows
```
{
  "streams": [
    {
      "stream": {
        "name": "cadence_memberships",
        "json_schema": {},
        "supported_sync_modes": ["full_refresh", "incremental"],
        "source_defined_cursor": true,
        "default_cursor_field": ["updated_at"],
        "source_defined_primary_key": [["id"]]
      },
      "sync_mode": "incremental",
      "destination_sync_mode": "overwrite",
      "cursor_field": ["updated_at"]
    }
  ]
}
```

Also you can look how the catalog is treated in the CDK's [read](https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/abstract_source.py#L107) method.

tuanchris replied:
I'm still not sure what you are suggesting. Is it not enough? https://github.com/airbytehq/airbyte/blob/22bf955d5bd17a9a405372a7df50267a23a4961d/airbyte-integrations/connectors/source-gcs/integration_tests/configured_catalog.json

tuanchris replied:
That makes sense. Thank you @davydov-d, I've implemented the logic now.

davydov-d commented on airbyte-integrations/connectors/source-gcs/source_gcs/helpers.py:
```diff
@@ -0,0 +1,56 @@
+
+
+def read_csv_file(blob: Blob, read_header_only=False):
+    file_obj = io.BytesIO()
```
please close the file-like object after having read from it

davydov-d reviewed (approved):
I've left one more comment, otherwise looks good to me

davydov-d commented:
@marcosmarxm please proceed with this PR when you have a chance

davydov-d commented:
> @marcosmarxm please proceed with this PR when you have a chance

cc @sh4sh can you pick up this one please? I guess @marcosmarxm is unavailable

marcosmarxm commented:
/test connector=connectors/source-gcs

marcosmarxm commented:
/publish connector=connectors/source-gcs

marcosmarxm reviewed (approved):
Thanks @tuanchris

marcosmarxm commented:
> renatodossantosleal

Can you open a feature request?
